src/std/text/json/output.ss

```diff
@@ -0,0 +1,202 @@
+                    (write-json-object val output env)
+                    (unless (null? rest) (write-char #\, output))
+                    (lp rest)))
+               ([]                      ; empty
+                (write-char #\} output)))))
+
+           (def (write-json-alist/sort alist output env)
+             (write-json-alist (json-sort-alist alist) output env))
+
+           (def (write-json-hash obj output env)
+             (def lst (hash->list obj))
+             (if (&env-sort-keys env)
+               (write-json-alist/sort lst output env)
+               (write-json-alist lst output env)))
+
+           (def (write-json-string obj output env)
+             (def escape
+               '((#\" . #\")
+                 (#\\ . #\\)
+                 (#\backspace . #\b)
+                 (#\newline . #\n)
+                 (#\page . #\f)
+                 (#\return . #\r)
+                 (#\tab . #\t)))
+
+             (def (printable-char? char)
+               (let (n (char->integer char))
+                 (cond
+                  ((##fx< n 32) #f)     ; unprintable ascii
+                  ((##fx= n 127) #f)    ; ascii DEL
+                  (else #t))))
+
+             (def (write-uchar char output)
+               (let (int (char->integer char))
+                 (write-string "\\u" output)
+                 (let lp ((n 0) (mask #xf000) (shift -12))
+                   (when (fx< n 4)
+                     (let (char (hex (arithmetic-shift (bitwise-and int mask) shift)))
+                       (write-char char output)
+                       (lp (fx+ n 1) (arithmetic-shift mask -4) (fx+ shift 4)))))))
+
+             (def (write-str obj output)
+               (let (len (string-length obj))
+                 (let lp ((n 0))
+                   (when (##fx< n len)
+                     (let (char (string-ref obj n))
+                       (cond
+                        ((assq char escape)
+                         => (lambda (esc)
+                              (write-char #\\ output)
+                              (write-char (cdr esc) output)))
+                        ((printable-char? char)
+                         (write-char char output))
+                        (else
+                         (write-uchar char output)))
+                       (lp (##fx+ n 1)))))))
+
+             (write-char #\" output)
+             (write-str obj output)
+             (write-char #\" output)))))))
+
+(defrule (writer-write-char char buf)
+  (&BufferedStringWriter-write-char-inline buf char))
+(defrule (writer-write-string str buf)
+  (&BufferedStringWriter-write-string buf str))
+
+(defrule (buffer-write-char char buf)
+  (&BufferedWriter-write-char-inline buf char))
+(defrule (buffer-write-string str buf)
+  (&BufferedWriter-write-string buf str))
+
+(defjson-writer port write-char write-string)
+(defjson-writer writer writer-write-char writer-write-string)
+(defjson-writer buffer buffer-write-char buffer-write-string)
```

Comment:
How much do we win by making it a macro vs a function with higher-order arguments?

Reply:
How about instead having "just" the buffered version, and then wrap port and writer with a buffered wrapper?

Reply:
Wrapping the port would be a msitake, we would double buffer.

Reply:
> How about instead having "just" the buffered version, and then wrap port and writer with a buffered wrapper?

In general that cant work fot input, the buffer might steal data.

Reply:
> How much do we win by making it a macro vs a function with higher-order arguments?

Quite a lot, inckuding arity checking ;)